pkg/services/datasources/service/datasource.go

```diff
@@ -252,6 +294,66 @@ func (s *Service) AddDataSource(ctx context.Context, cmd *datasources.AddDataSou
 	})
 }
 
+// This will valid validate the instance settings and mutate the cmd with the processed values
+func (s *Service) mutate(ctx context.Context, pluginContext backend.PluginContext, settings *backend.DataSourceInstanceSettings) (*backend.DataSourceInstanceSettings, error) {
+	operation := backend.AdmissionRequestCREATE
+
+	if len(settings.Name) > maxDatasourceNameLen {
+		return nil, datasources.ErrDataSourceNameInvalid.Errorf("max length is %d", maxDatasourceNameLen)
+	}
+
+	if len(settings.URL) > maxDatasourceUrlLen {
+		return nil, datasources.ErrDataSourceURLInvalid.Errorf("max length is %d", maxDatasourceUrlLen)
+	}
+
+	if settings.Type == "" {
+		return settings, nil // NOOP for tests
+	}
+
+	// Make sure it is a known plugin type
+	p, found := s.pluginStore.Plugin(ctx, settings.Type)
+	if !found {
+		return nil, fmt.Errorf("plugin '%s' not found", settings.Type)
+	}
+
+	// When the APIVersion is set, the client must also implement ProcessInstanceSettings
+	if p.APIVersion == "" {
+		if settings.APIVersion != "" {
+			return nil, fmt.Errorf("invalid request apiVersion (datasource does not have one configured)")
+		}
+		return settings, nil // NOOP
+	}
+
+	req := &backend.AdmissionRequest{
+		Operation:     backend.AdmissionRequestCREATE,
+		PluginContext: pluginContext,
+		Kind:          settings.GVK(),
+		ObjectBytes:   settings.ProtoBytes(),
+	}
+
+	// Set the old bytes and change the operation
+	if pluginContext.DataSourceInstanceSettings != nil {
+		req.Operation = backend.AdmissionRequestUPDATE
+		req.OldObjectBytes = pluginContext.DataSourceInstanceSettings.ProtoBytes()
+	}
+
+	rsp, err := s.pluginClient.MutateAdmission(ctx, req)
+	if err != nil {
+		if errors.Is(err, plugins.ErrMethodNotImplemented) {
```

Comment:
Added here so that we see the error "plugin (%s) with apiVersion=%s must implement MutateAdmission" rather than just not implemented

Reply:
Looks like this still could live in plugin client and return a proper errutil.Error.

Reply:
ya -- but I vote we tackle better error handling in another PR.  In particular converting the result (a raw k8s style error!) into errutil needs some attention since we can not set an error code